### PR TITLE
Avoid calling ReportErrors concurrently from multiple threads

### LINKED ISSUE #12550: ReportErrors is not thread safe but can be invoked concurrently with /parallel
For the most part, calls to`CommonCompiler.ReportErrors` do not happen concurrently. However, there is one call in each of VB and C#'s `ParseFile` when there is an I/O error that can be invoked concurrently in the `/parallel` case. There are two unsafe operations with invoking `CommonCompiler.ReportErrors` concurrently:

1. Updating `HashSet<Diagnostic> _reportedDiagnostics`,
2. Calling `ErrorLogger.LogDiagnostic` for `/errorlog`

So far, I have only managed to construct a repro (detailed below) that illustrates (2).

**Version Used**: 1.3.1.60616 (still present in 2.0 by code inspection of master)

**Steps to Reproduce**:

1.  Create two empty files: a.cs and b.cs
2.  Change permissions of both files to deny all access
3.  `csc a.cs b.cs /parallel /errorlog:log.json`
4. Repeat 1-3 several times

**Expected Behavior**:
A valid error log is produced.

**Actual Behavior**:
It sometimes works (concurrency issue), and sometimes fails in at least two different ways: 

1. A malformed error log is silently produced.
2. An unhandled exception is raised:
```
Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.CodeAnalysis.ErrorLogger.DiagnosticDescriptorSet.Add(DiagnosticDescriptor descriptor)
   at Microsoft.CodeAnalysis.ErrorLogger.LogDiagnostic(Diagnostic diagnostic)
   at Microsoft.CodeAnalysis.CommonCompiler.ReportErrors(IEnumerable`1 diagnostics, TextWriter consoleOutput, ErrorLogger errorLoggerOpt)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompiler.ParseFile(TextWriter consoleOutput, CSharpParseOptions parseOptions, CSharpParseOptions scriptParseOptions, Boolean& hadErrors, CommandLineSourceFile file, ErrorLogger errorLogger, String& normalizedFilePath)
```

### PR DESCRIPTION
**Pull request by agocke**
ReportErrors is not safe to call from multiple threads.

Fixes #12550.

### CODE REVIEW

**Comment by agocke** [comment]
@jaredpar Should I rebase this to a different branch for ask mode?

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs [hunk 1]
```diff
@@ -44,29 +44,35 @@ public override Compilation CreateCompilation(TextWriter consoleOutput, TouchedF
...
 
             // If errors had been reported in ParseFile, while trying to read files, then we should simply exit.
             if (hadErrors)
             {
+                ReportErrors(diagnosticBag.ToReadOnlyAndFree(), consoleOutput, errorLogger);
                 return null;
             }
+            else
+            {
+                diagnosticBag.Free();
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs]: This appears to discard diagnostics if there were only warnings.

Comment (agocke) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs]: Actually, I suppose any level diagnostic could be added to the bag as long as `content` is null. And no matter what the level of diagnostic we will report it -- `hadErrors` is maybe improperly named. I'll also rename that variable.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs [hunk 2]
```diff
@@ -44,29 +44,35 @@ public override Compilation CreateCompilation(TextWriter consoleOutput, TouchedF
...
             }
 
             // If errors had been reported in ParseFile, while trying to read files, then we should simply exit.
             if (hadErrors)
             {
+                ReportErrors(diagnosticBag.ToReadOnlyAndFree(), consoleOutput, errorLogger);
                 return null;
             }
+            else
+            {
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs]: Should we assert the bag is empty here?

Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs]: Or at least that all remaining diagnostics are hidden?

**Comment by agocke** [comment]
@gafter I think I've addressed your concerns

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs [hunk 3]
```diff
@@ -43,30 +43,38 @@ public override Compilation CreateCompilation(TextWriter consoleOutput, TouchedF
...
             // If errors had been reported in ParseFile, while trying to read files, then we should simply exit.
             if (hadErrors)
             {
+                Debug.Assert(diagnosticBag.HasAnyErrors());
+                ReportErrors(diagnosticBag.ToReadOnlyAndFree(), consoleOutput, errorLogger);
                 return null;
             }
+            else
+            {
+                Debug.Assert(!diagnosticBag.HasAnyErrors());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs]: >diagnosticBag.HasAnyErrors() [](start = 30, length = 28)

Shouldn't we assert that the bag is empty?

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs (hunk 3 above)
Comment (agocke) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCompiler.cs]: @AlekseyTs Never mind, I just found the IsEmptyWithoutResolution property. I'm blind, apparently.

**Pull request merged**